File: packages/vtx-parser/python/vtx_parser/clock_drift.py
"""
Clock drift analysis for VTX v1.2 sync records.

Sync records are observations of the device time base, never corrections
applied to it. This module turns them into a drift rate, a correction
function, and a quality metric that tells a caller whether the fit is
trustworthy enough to use.

Mirrors packages/vtx-parser/src/clock-drift.ts. See
packages/vtx-format/spec/v1.2-clock-sync.md for the record layout.
"""

from dataclasses import dataclass, field
from statistics import median
from typing import Callable, List, Optional, Tuple

from .types import ClockSyncRecord


@dataclass
class ClockSyncObservation:
    """A sync record reduced to a single (device time, clock offset) point."""
    device_ms: float   # device millis() at the midpoint of the exchange
    offset_ms: float   # phone-minus-device offset, transport delay cancelled
    rtt_ms: float      # round trip, phone processing removed
    index: int         # index of the source record


@dataclass
class RejectedObservation:
    observation: ClockSyncObservation
    reason: str


@dataclass
class ClockDriftFit:
    """Result of fitting drift to a set of sync records."""

    ppm: float
    """Drift rate in parts per million. Positive means the device clock runs
    slow relative to the phone (millis() advances less than true time)."""

    offset_at_t0_ms: float
    t0_device_ms: float
    used: List[ClockSyncObservation]
    rejected: List[RejectedObservation]
    residual_rms_ms: float
    residual_max_ms: float
    span_ms: float
    trustworthy: bool
    warnings: List[str]
    correct: Callable[[float], float]
    """Map device millis() -> estimated true unix ms. Always defined; degrades
    to a constant offset with one record and to identity with none."""

    suspected_steps: List[int] = field(default_factory=list)


def to_observation(record: ClockSyncRecord, index: int = 0) -> ClockSyncObservation:
    """
    Reduce a sync record to an offset/RTT observation using the NTP estimator.

        offset = ((t2 - t1) + (t3 - t4)) / 2
        rtt    = (t4 - t1) - (t3 - t2)

    t1/t4 are device millis(); t2/t3 are phone unix ms. The two clocks have
    different epochs, so `offset` is the epoch difference — exactly the
    quantity whose drift we want.
    """
    t1 = record.t1_device_ms
    t4 = record.t4_device_ms
    t2 = record.t2_phone_unix_ms
    t3 = record.t3_phone_unix_ms

    offset_ms = ((t2 - t1) + (t3 - t4)) / 2.0
    rtt_ms = (t4 - t1) - (t3 - t2)
    return ClockSyncObservation(
        device_ms=(t1 + t4) / 2.0,
        offset_ms=offset_ms,
        rtt_ms=rtt_ms,
        index=index,
    )


def detect_clock_steps(
    obs: List[ClockSyncObservation], step_threshold_ms: float = 250.0
) -> List[int]:
    """
    Flag observations whose offset jumps discontinuously relative to their
    neighbours — the signature of the phone's own clock stepping (an NTP
    correction mid-ride) rather than the device crystal drifting.

    Crystal drift is slow and monotonic: at 50ppm the offset moves 3ms per
    minute. A phone NTP correction moves it tens or hundreds of ms between
    adjacent samples.

    With fewer than 3 observations a step is indistinguishable from drift and
    nothing is flagged.
    """
    if len(obs) < 3:
        return []

    slopes = []
    for i in range(1, len(obs)):
        dt = obs[i].device_ms - obs[i - 1].device_ms
        if dt <= 0:
            continue
        slopes.append((obs[i].offset_ms - obs[i - 1].offset_ms) / dt)

    if not slopes:
        return []
    med_slope = median(slopes)

    steps: List[int] = []
    for i in range(1, len(obs)):
        dt = obs[i].device_ms - obs[i - 1].device_ms
        if dt <= 0:
            continue
        predicted = obs[i - 1].offset_ms + med_slope * dt
        if abs(obs[i].offset_ms - predicted) > step_threshold_ms:
            steps.append(i)
    return steps


def compute_clock_drift(
    records: List[ClockSyncRecord],
    max_rtt_ms: float = 500.0,
    rtt_outlier_factor: float = 4.0,
    step_threshold_ms: float = 250.0,
) -> ClockDriftFit:
    """
    Fit clock drift from sync records.

    Method: ordinary least squares of offset against device time, after
    rejecting high-RTT observations.

    Why linear. Crystal frequency error is dominated by a fixed manufacturing
    offset plus a temperature coefficient. Over a ride the manufacturing term
    is constant, so offset accumulates linearly in time and a straight line is
    the correct model. The residual tells you when that assumption fails.

    Where it breaks:
      - Thermal transients. A tuning-fork crystal's frequency follows a
        parabola in temperature (about -0.035 ppm/degC^2 around a ~25C
        turnover). A device carried from indoors into cold air sweeps that
        curve and the drift rate is genuinely not constant. This appears as
        structured residuals, which is why residual_rms_ms is reported.
      - A phone clock step. NTP correcting the phone mid-ride moves the offset
        discontinuously; one OLS line absorbs it as slope error across the
        whole recording. detect_clock_steps() flags it — refit per segment.
      - Short spans. Over 60s, BLE latency noise of tens of ms dominates a
        drift of microseconds.

    Degenerate inputs never raise:
      - zero records -> ppm 0, correct() is identity, untrustworthy
      - one record   -> constant offset, no rate, untrustworthy
    """
    all_obs = [to_observation(r, i) for i, r in enumerate(records)]
    all_obs.sort(key=lambda o: o.device_ms)

    rejected: List[RejectedObservation] = []
    warnings: List[str] = []

    def finite(x: float) -> bool:
        return x == x and x not in (float("inf"), float("-inf"))

    candidates: List[ClockSyncObservation] = []
    for o in all_obs:
        if not (finite(o.offset_ms) and finite(o.rtt_ms)):
            rejected.append(RejectedObservation(o, "non-finite timestamps"))
            continue
        # Negative RTT is physically impossible: the phone claimed a longer
        # processing time than the whole device-observed round trip.
        if o.rtt_ms < 0:
            rejected.append(RejectedObservation(o, "negative RTT"))
            continue
        if o.rtt_ms > max_rtt_ms:
            rejected.append(
                RejectedObservation(o, f"RTT {o.rtt_ms:.0f}ms > {max_rtt_ms:.0f}ms")
            )
            continue
        candidates.append(o)

    # Relative RTT filter, once a median means something.
    if len(candidates) >= 4:
        med_rtt = median([o.rtt_ms for o in candidates])
        # Only meaningful if the median is non-trivial; with a sub-ms median
        # every real sample would trip a 4x rule.
        if med_rtt >= 5:
            limit = med_rtt * rtt_outlier_factor
            kept = []
            for o in candidates:
                if o.rtt_ms > limit:
                    rejected.append(
                        RejectedObservation(
                            o,
                            f"RTT {o.rtt_ms:.0f}ms > {rtt_outlier_factor}x "
                            f"median ({med_rtt:.0f}ms)",
                        )
                    )
                else:
                    kept.append(o)
            candidates = kept

    suspected_steps = detect_clock_steps(candidates, step_threshold_ms)

    # --- degenerate cases ---
    if not candidates:
        warnings.append(
            "no sync records: device clock cannot be characterized"
            if not all_obs
            else "all sync records rejected as unusable"
        )
        return ClockDriftFit(
            ppm=0.0,
            offset_at_t0_ms=0.0,
            t0_device_ms=0.0,
            used=[],
            rejected=rejected,
            residual_rms_ms=0.0,
            residual_max_ms=0.0,
            span_ms=0.0,
            trustworthy=False,
            warnings=warnings,
            correct=lambda device_ms: device_ms,
            suspected_steps=[],
        )

    if len(candidates) == 1:
        only = candidates[0]
        warnings.append("single sync record: offset known, drift rate unknowable")
        return ClockDriftFit(
            ppm=0.0,
            offset_at_t0_ms=only.offset_ms,
            t0_device_ms=only.device_ms,
            used=candidates,
            rejected=rejected,
            residual_rms_ms=0.0,
            residual_max_ms=0.0,
            span_ms=0.0,
            trustworthy=False,
            warnings=warnings,
            # Best available: a constant offset, exactly what the pre-v1.2
            # single CMD_SYNC_CLOCK gave us.
            correct=lambda device_ms, _o=only.offset_ms: device_ms + _o,
            suspected_steps=[],
        )

    # --- ordinary least squares ---
    t0 = candidates[0].device_ms
    n = len(candidates)
    mx = sum(o.device_ms - t0 for o in candidates) / n
    my = sum(o.offset_ms for o in candidates) / n

    sxy = sum((o.device_ms - t0 - mx) * (o.offset_ms - my) for o in candidates)
    sxx = sum((o.device_ms - t0 - mx) ** 2 for o in candidates)

    span_ms = candidates[-1].device_ms - candidates[0].device_ms

    # slope is ms of offset per ms of device time — dimensionless, so ppm is a
    # direct 1e6 scaling.
    slope = (sxy / sxx) if sxx > 0 else 0.0
    intercept = my - slope * mx

    if sxx == 0:
        warnings.append(
            "all sync records share one device timestamp: no time base to fit"
        )

    sum_sq = 0.0
    max_abs = 0.0
    for o in candidates:
        predicted = intercept + slope * (o.device_ms - t0)
        resid = o.offset_ms - predicted
        sum_sq += resid * resid
        max_abs = max(max_abs, abs(resid))
    residual_rms_ms = (sum_sq / n) ** 0.5

    # --- trust assessment ---
    if n < 3:
        warnings.append(
            f"only {n} usable sync records: drift rate is poorly constrained"
        )
    if span_ms < 10 * 60 * 1000:
        warnings.append(
            f"sync span {span_ms / 60000:.1f} min is short: BLE latency noise "
            "dominates the drift signal"
        )
    if residual_rms_ms > 50:
        warnings.append(
            f"residual RMS {residual_rms_ms:.1f}ms is large: linear drift may "
            "not hold (thermal transient or clock step)"
        )
    if suspected_steps:
        warnings.append(
            f"{len(suspected_steps)} suspected clock step(s) at observation "
            f"index {', '.join(str(i) for i in suspected_steps)}: a single "
            "line will absorb the step as slope error"
        )

    ppm = slope * 1e6
    if abs(ppm) > 200:
        warnings.append(
            f"fitted drift {ppm:.1f}ppm exceeds plausible crystal tolerance: "
            "suspect a clock step or bad samples"
        )

    return ClockDriftFit(
        ppm=ppm,
        offset_at_t0_ms=intercept,
        t0_device_ms=t0,
        used=candidates,
        rejected=rejected,
        residual_rms_ms=residual_rms_ms,
        residual_max_ms=max_abs,
        span_ms=span_ms,
        trustworthy=not warnings,
        warnings=warnings,
        correct=(
            lambda device_ms, _i=intercept, _s=slope, _t=t0: device_ms
            + _i
            + _s * (device_ms - _t)
        ),
        suspected_steps=suspected_steps,
    )

File: packages/vtx-parser/src/clock-drift.ts
/**
 * Clock drift analysis for VTX v1.2 sync records.
 *
 * Sync records are observations of the device time base, never corrections
 * applied to it. This module turns them into a drift rate, a correction
 * function, and a quality metric that tells a caller whether the fit is
 * trustworthy enough to use.
 *
 * See packages/vtx-format/spec/v1.2-clock-sync.md for the record layout.
 */

import { ClockSyncRecord } from './types';

/** A sync record reduced to a single (device time, clock offset) observation. */
export interface ClockSyncObservation {
  /** Device millis() at the midpoint of the exchange */
  deviceMs: number;
  /** Phone-minus-device clock offset in ms, transport delay cancelled */
  offsetMs: number;
  /** Round-trip time in ms, phone processing removed */
  rttMs: number;
  /** Index of the source record in the input array */
  index: number;
}

export interface ClockDriftOptions {
  /**
   * Discard observations whose RTT exceeds this many ms.
   * Default 500. A BLE exchange at a 30-50ms connection interval normally
   * completes in well under 200ms; anything far above that is a stalled
   * radio, not a measurement.
   */
  maxRttMs?: number;
  /**
   * Discard observations whose RTT exceeds this multiple of the median RTT.
   * Default 4. Catches the case where the whole link is slow and the fixed
   * threshold above is too permissive.
   */
  rttOutlierFactor?: number;
  /**
   * Residual threshold (ms) above which an observation is treated as a
   * step rather than drift. Default 250.
   */
  stepThresholdMs?: number;
}

export interface ClockDriftFit {
  /**
   * Measured drift rate in parts per million. Positive means the device
   * clock runs slow relative to the phone (device millis() advances less
   * than true elapsed time).
   */
  ppm: number;
  /** Fitted offset (ms) at deviceMs = t0 */
  offsetAtT0Ms: number;
  /** Reference device time for the fit (first retained observation) */
  t0DeviceMs: number;
  /** Observations retained after filtering */
  used: ClockSyncObservation[];
  /** Observations discarded, with the reason */
  rejected: Array<ClockSyncObservation & { reason: string }>;
  /** Root-mean-square residual (ms) of retained observations about the fit */
  residualRmsMs: number;
  /** Largest absolute residual (ms) */
  residualMaxMs: number;
  /** Span of retained observations in ms of device time */
  spanMs: number;
  /**
   * Whether the fit is trustworthy. False when there are too few points,
   * too short a span, or residuals large enough to suggest the linear
   * model does not hold (a phone clock step, most likely).
   */
  trustworthy: boolean;
  /** Human-readable reasons the fit is not trustworthy (empty if it is) */
  warnings: string[];
  /**
   * Map a device millis() value to estimated true unix ms.
   * Always defined — degrades to the best available estimate, and to the
   * identity-with-offset when only one observation exists.
   */
  correct: (deviceMs: number) => number;
  /**
   * Indices of observations that look like a discontinuity (phone clock
   * step) rather than drift. See detectClockSteps().
   */
  suspectedSteps: number[];
}

/**
 * Reduce a sync record to an offset/RTT observation using the NTP estimator.
 *
 *   offset = ((t2 - t1) + (t3 - t4)) / 2
 *   rtt    = (t4 - t1) - (t3 - t2)
 *
 * t1/t4 are device millis(); t2/t3 are phone unix ms. The two clocks have
 * different epochs, so `offset` is the epoch difference — that is exactly
 * the quantity whose drift we want.
 */
export function toObservation(
  record: ClockSyncRecord,
  index: number = 0
): ClockSyncObservation {
  const { t1DeviceMs, t2PhoneUnixMs, t3PhoneUnixMs, t4DeviceMs } = record;
  const offsetMs =
    (t2PhoneUnixMs - t1DeviceMs + (t3PhoneUnixMs - t4DeviceMs)) / 2;
  const rttMs = t4DeviceMs - t1DeviceMs - (t3PhoneUnixMs - t2PhoneUnixMs);
  return {
    deviceMs: (t1DeviceMs + t4DeviceMs) / 2,
    offsetMs,
    rttMs,
    index,
  };
}

function median(values: number[]): number {
  if (values.length === 0) return NaN;
  const s = [...values].sort((a, b) => a - b);
  const mid = s.length >> 1;
  return s.length % 2 ? s[mid] : (s[mid - 1] + s[mid]) / 2;
}

/**
 * Flag observations whose offset jumps discontinuously relative to their
 * neighbours — the signature of the phone's own clock stepping (an NTP
 * correction mid-ride), as opposed to the device crystal drifting.
 *
 * Crystal drift is slow and monotonic: at 50ppm the offset moves 3ms per
 * minute. A phone NTP correction moves it by tens or hundreds of ms between
 * two adjacent samples. This looks at the first difference of the offset
 * series and flags points where it exceeds `stepThresholdMs` above what the
 * surrounding cadence would predict.
 *
 * Returns indices into the observation array. With fewer than 3 observations
 * a step is indistinguishable from drift and nothing is flagged.
 */
export function detectClockSteps(
  obs: ClockSyncObservation[],
  stepThresholdMs: number = 250
): number[] {
  if (obs.length < 3) return [];

  const steps: number[] = [];
  // Per-ms drift implied by each adjacent pair
  const slopes: number[] = [];
  for (let i = 1; i < obs.length; i++) {
    const dt = obs[i].deviceMs - obs[i - 1].deviceMs;
    if (dt <= 0) continue;
    slopes.push((obs[i].offsetMs - obs[i - 1].offsetMs) / dt);
  }
  const medSlope = median(slopes);

  for (let i = 1; i < obs.length; i++) {
    const dt = obs[i].deviceMs - obs[i - 1].deviceMs;
    if (dt <= 0) continue;
    const predicted = obs[i - 1].offsetMs + medSlope * dt;
    if (Math.abs(obs[i].offsetMs - predicted) > stepThresholdMs) {
      steps.push(i);
    }
  }
  return steps;
}

/**
 * Fit clock drift from sync records.
 *
 * Method: ordinary least squares of offset against device time, after
 * rejecting high-RTT observations.
 *
 * Why linear. Crystal frequency error is dominated by a fixed manufacturing
 * offset plus a temperature coefficient. Over a ride the manufacturing term
 * is constant, so the offset accumulates linearly in time and a straight
 * line is the correct model. The residual tells you when that assumption
 * fails.
 *
 * Where it breaks:
 *  - **Thermal transients.** A tuning-fork crystal's frequency follows a
 *    parabola in temperature (roughly -0.035 ppm/degC^2 about a ~25C turnover).
 *    A device that starts indoors and is ridden into cold air sweeps that
 *    curve, and drift rate is genuinely not constant. This shows up as
 *    structured — not random — residuals, which is why residualRmsMs is
 *    reported rather than just an R-squared.
 *  - **A phone clock step.** NTP correcting the phone mid-ride moves the
 *    offset discontinuously. A single OLS line absorbs it as a slope error
 *    across the whole recording. detectClockSteps() flags this; the caller
 *    should refit per-segment rather than trusting one line.
 *  - **Short spans.** Over 60s of observations, BLE latency noise of tens of
 *    ms dominates a drift of microseconds. The fit is reported but marked
 *    untrustworthy.
 *
 * Degenerate inputs never throw:
 *  - zero records -> ppm 0, correct() is identity on the device base, untrustworthy
 *  - one record   -> constant offset, no rate, untrustworthy
 */
export function computeClockDrift(
  records: ClockSyncRecord[],
  options: ClockDriftOptions = {}
): ClockDriftFit {
  const maxRttMs = options.maxRttMs ?? 500;
  const rttOutlierFactor = options.rttOutlierFactor ?? 4;
  const stepThresholdMs = options.stepThresholdMs ?? 250;

  const all = records.map((r, i) => toObservation(r, i));
  all.sort((a, b) => a.deviceMs - b.deviceMs);

  const rejected: Array<ClockSyncObservation & { reason: string }> = [];
  const warnings: string[] = [];

  // Negative RTT is physically impossible: the phone reported a longer
  // processing time than the whole device-observed round trip. Clock
  // granularity or a misbehaving responder.
  let candidates = all.filter((o) => {
    if (!Number.isFinite(o.offsetMs) || !Number.isFinite(o.rttMs)) {
      rejected.push({ ...o, reason: 'non-finite timestamps' });
      return false;
    }
    if (o.rttMs < 0) {
      rejected.push({ ...o, reason: 'negative RTT' });
      return false;
    }
    if (o.rttMs > maxRttMs) {
      rejected.push({ ...o, reason: `RTT ${o.rttMs.toFixed(0)}ms > ${maxRttMs}ms` });
      return false;
    }
    return true;
  });

  // Relative RTT filter, once there are enough points for a median to mean
  // anything.
  if (candidates.length >= 4) {
    const medRtt = median(candidates.map((o) => o.rttMs));
    const limit = medRtt * rttOutlierFactor;
    // Only meaningful if the median is non-trivial; with a sub-ms median
    // every real sample would trip a 4x rule.
    if (medRtt >= 5) {
      candidates = candidates.filter((o) => {
        if (o.rttMs > limit) {
          rejected.push({
            ...o,
            reason: `RTT ${o.rttMs.toFixed(0)}ms > ${rttOutlierFactor}x median (${medRtt.toFixed(0)}ms)`,
          });
          return false;
        }
        return true;
      });
    }
  }

  const suspectedSteps = detectClockSteps(candidates, stepThresholdMs);

  // --- degenerate cases ---
  if (candidates.length === 0) {
    warnings.push(
      all.length === 0
        ? 'no sync records: device clock cannot be characterized'
        : 'all sync records rejected as unusable'
    );
    return {
      ppm: 0,
      offsetAtT0Ms: 0,
      t0DeviceMs: 0,
      used: [],
      rejected,
      residualRmsMs: 0,
      residualMaxMs: 0,
      spanMs: 0,
      trustworthy: false,
      warnings,
      // No information: pass device time through unchanged.
      correct: (deviceMs: number) => deviceMs,
      suspectedSteps: [],
    };
  }

  if (candidates.length === 1) {
    const only = candidates[0];
    warnings.push('single sync record: offset known, drift rate unknowable');
    return {
      ppm: 0,
      offsetAtT0Ms: only.offsetMs,
      t0DeviceMs: only.deviceMs,
      used: candidates,
      rejected,
      residualRmsMs: 0,
      residualMaxMs: 0,
      spanMs: 0,
      trustworthy: false,
      warnings,
      // Best available: a constant offset, exactly what the pre-v1.2
      // single CMD_SYNC_CLOCK gave us.
      correct: (deviceMs: number) => deviceMs + only.offsetMs,
      suspectedSteps: [],
    };
  }

  // --- ordinary least squares ---
  const t0 = candidates[0].deviceMs;
  const n = candidates.length;
  let sx = 0;
  let sy = 0;
  for (const o of candidates) {
    sx += o.deviceMs - t0;
    sy += o.offsetMs;
  }
  const mx = sx / n;
  const my = sy / n;

  let sxy = 0;
  let sxx = 0;
  for (const o of candidates) {
    const dx = o.deviceMs - t0 - mx;
    sxy += dx * (o.offsetMs - my);
    sxx += dx * dx;
  }

  const spanMs = candidates[n - 1].deviceMs - candidates[0].deviceMs;

  // slope is ms of offset per ms of device time — dimensionless, so ppm is
  // a direct 1e6 scaling.
  const slope = sxx > 0 ? sxy / sxx : 0;
  const intercept = my - slope * mx;

  if (sxx === 0) {
    warnings.push('all sync records share one device timestamp: no time base to fit');
  }

  let sumSq = 0;
  let maxAbs = 0;
  for (const o of candidates) {
    const predicted = intercept + slope * (o.deviceMs - t0);
    const resid = o.offsetMs - predicted;
    sumSq += resid * resid;
    maxAbs = Math.max(maxAbs, Math.abs(resid));
  }
  const residualRmsMs = Math.sqrt(sumSq / n);

  // --- trust assessment ---
  if (n < 3) {
    warnings.push(`only ${n} usable sync records: drift rate is poorly constrained`);
  }
  if (spanMs < 10 * 60 * 1000) {
    warnings.push(
      `sync span ${(spanMs / 60000).toFixed(1)} min is short: BLE latency noise dominates the drift signal`
    );
  }
  if (residualRmsMs > 50) {
    warnings.push(
      `residual RMS ${residualRmsMs.toFixed(1)}ms is large: linear drift may not hold (thermal transient or clock step)`
    );
  }
  if (suspectedSteps.length > 0) {
    warnings.push(
      `${suspectedSteps.length} suspected clock step(s) at observation index ${suspectedSteps.join(', ')}: a single line will absorb the step as slope error`
    );
  }

  const ppm = slope * 1e6;
  if (Math.abs(ppm) > 200) {
    warnings.push(
      `fitted drift ${ppm.toFixed(1)}ppm exceeds plausible crystal tolerance: suspect a clock step or bad samples`
    );
  }

  return {
    ppm,
    offsetAtT0Ms: intercept,
    t0DeviceMs: t0,
    used: candidates,
    rejected,
    residualRmsMs,
    residualMaxMs: maxAbs,
    spanMs,
    trustworthy: warnings.length === 0,
    warnings,
    correct: (deviceMs: number) =>
      deviceMs + intercept + slope * (deviceMs - t0),
    suspectedSteps,
  };
}
